### ISSUE TITLE
Add configuration option to not emit 'unreachable code' hints on python version checks

### ISSUE THREAD

**Issue by KotlinIsland** [microsoft/pyright#2258]
```py
if sys.version_info < (3, 9):
    print("Ancient python version detected 😳😳😳")  # unreachable code detected!!!!!!!!
elif sys.version_info < (3, 11):
    print("standard python version detected 😒")   # unreachable code detected!!!!!!!!
else:
    print("based 😎")
```
I don't want to see these unreachable code hints when I'm writing code, makes me think I've made an error (as unreachable code usually indicates a logic error)

pyright should have an option like "showUnreachableCodeHintsOnVersionChecks" and it should default to false. 

Maybe it could be specified with a range somehow `>=3.9, <3.12` (as in, this project is compatible with python 3.9 - 3.11, so only show unreachable hints when a different version is checked)

If you ever have an explicit version check in your codebase, why would you ever want to see it marked as bad?

**Comment by erictraut** [comment]
Which editor are you using, and how is it displaying these hints? Hints are meant to be displayed as a subtle "grayed out" appearance. Here's a screen shot in VS Code.

![Screen Shot 2021-09-02 at 8 58 00 PM](https://user-images.githubusercontent.com/7040122/131948507-8e207f0c-c827-4e90-a5f3-bb28ea55ca03.png)

If your editor is displaying these hints as regular diagnostics, I would consider that a bug, and I'd recommend reporting that to the maintainer of your editor or the LSP plugin.

Some LSP plugins allow you to disable these hints if you don't like the way they are displayed.

**Comment by erictraut** [comment]
I'm going to close this because I don't think it makes sense for pyright to add new settings to work around incorrect behaviors in a specific editor.

Please work with the maintainer of your editor or LSP plugin and encourage them to interpret and display hints correctly. If that's not practical, they can indicate to pyright that they don't support hints, and pyright won't generate them. If the editor indicates that it supports hints, it should display them properly.

**Comment by KotlinIsland** [comment]
```py
import sys
from typing import NoReturn

if sys.version_info < (8,):
    print("huh?")  # actual defect, no warning

if sys.version_info < (3, 8):
    print("huh?")  # not a defect, just a normal version check

def foo():
    return
    print("huh?")  # actual defect, syntacticly unreachable code

def bar() -> NoReturn:
    sys.exit(1)

def baz():
    bar()
    print("huh?")  # actual defect, type-based unreachable code

```
![image](https://user-images.githubusercontent.com/65446343/132118420-fd920e1c-d0ea-4dd7-8c97-6fad048e4401.png)

The issue is that pyright emits the same 'unreachable code' for real problems and 100% intentional and correct logic.

This is a real issue that pyright is reporting a false positive for valid code.

In my opinion:
'unreachable code' should not be reported for python version checks, I understand that the content of these blocks won't be checked by pyright, but marking it in the exact same way as real unreachable code is extremely misleading and confusing.

**Comment by fischercer** [comment]
I have a very similar issue. But instead of the Python version this hits when using `sys.platform`. So when I have to execute some code only on Windows I will get unreachable code errors for this part of the code when editing it on Linux and vice versa. 
Is there a workaround in disable this?

**Comment by KotlinIsland** [comment]
"pythonPlatform"
 https://github.com/microsoft/pyright/blob/main/docs/configuration.md#:~:text=one%20is%20present.-,pythonPlatform,-%5Bstring%2C%20optional%5D%3A%20Specifies

**Comment by davesque** [comment]
I have to disagree that this is an issue with code tools not handling errors properly.  Part of the issue is that there appears to be no way to squelch the message.  I believe `# type: ignore` doesn't work in this case.  If you're managing an open source project, it gets pretty annoying to have to parse a list of known errors in the output of your linting tools.  Supporting multiple python versions with a package is such a common thing that I feel there should be specific functionality to accommodate this in pyright.  It surely broadens the manner in which pyright decides if code is "correct", but it seems like a pretty high value feature.  If there's a way to silence these errors that I'm not aware of, I would be interested to know.  I would probably care a lot less about the issue if that were the case.

**Comment by erictraut** [comment]
If you run pyright as a command-line tool, it never emits any "unreachable code" hints. If you run pyright as a language server, it emits "unreachable code" hints _only if the client requests them_. It is then up to the client to use these hints appropriately. For example, VS Code displays unreachable code using a subtle grayed-out coloring scheme, and it does not treat these hints like other diagnostics. If your editor is requesting "unreachable code" hints and then treating them as a general diagnostic, I would consider that a bug — or a poor design choice.

If a client request "unreachable code" hints from the language server, it should use these hints appropriately.

**Comment by froh** [comment]
Hi,

in an active development project of mine I want to force myself to refactor some code to newer type checking once I can roll out python 3.12

```python
if sys.version_info >= (3, 12):
    raise SyntaxError("refactor to py 3.12 generic type alias")

    # refactor like so:

    #     type NestedDirectory[_T] = dict[str, NestedDirectory[_T] | _T | None]
    #     class Item[_T](int): ...

# else:
_T = TypeVar("_T")
_U = TypeVar("_U")

# end if...

```

alas, any and all code below the raise will be marked as unreachable.

too bad, so I have to _remember_

or rewrite the version check like so

```python

if (V := sys.version_info).major == 3 and V.minor >= 12 or V.major > 3:
    raise

```

just to confuse pyright?  ugly.
